### ISSUE TITLE
Keys without UID are ignored

### ISSUE THREAD

**Issue by heiglandreas** [phar-io/phive#281]
When trying to install a phar that was signed with a key where the public key is on `keys.openpgp.net` the result will always be `[WARNING]  Parsing key data failed with error code 0: No UIDs in key found` and the key will not be used. As that keyserver always removes the UID from the key that is to be expected...

How can we circumvent this issue?

**Comment by theseer** [comment]
Your questions and this issue confuse me ;)

Firstly, I assume you mean `keys.openpgp.org` rather then `.net`, right? Because I wasn't aware of a keyserver under openpgp.net. Trying to connect to that server times out for me...

Secondly, removing the `UID` fields kills the very idea of a keyserver as it makes the public key basically useless, so I wouldn't know why anyone would want that?

I also cannot reproduce this:

```
theseer@nyda /tmp/x5 $ wget -O 0x94D5D637E6659789.asc "https://keys.openpgp.org/pks/lookup/?op=get&options=mr&search=0x94D5D637E6659789"
--2020-09-24 15:28:08--  https://keys.openpgp.org/pks/lookup/?op=get&options=mr&search=0x94D5D637E6659789
Resolving keys.openpgp.org (keys.openpgp.org)... 37.218.245.50, 2a00:c6c0:0:154:1::1
Connecting to keys.openpgp.org (keys.openpgp.org)|37.218.245.50|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7754 (7,6K) [application/pgp-keys]
Saving to: ‘0x94D5D637E6659789.asc’

0x94D5D637E6659789.asc                     100%[======================================================================================>]   7,57K  --.-KB/s    in 0s      

2020-09-24 15:28:09 (132 MB/s) - ‘0x94D5D637E6659789.asc’ saved [7754/7754]
```

---

```
theseer@nyda /tmp/x5 $ gpg --homedir /tmp/x5 --quiet --with-colons --no-permission-warning --import 0x94D5D637E6659789.asc 
```

---

```
theseer@nyda /tmp/x5 $ gpg --homedir /tmp/x5 --quiet --with-colons --list-keys --no-permission-warning --with-fingerprint --with-fingerprint --fixed-list-mode  0x94D5D637E6659789
tru::1:1600954153:0:3:1:5
pub:-:4096:1:94D5D637E6659789:1599119968:1914660000::-:::cESCA::::::23::0:
fpr:::::::::9AFB471ED245B8BC56CB1EE694D5D637E6659789:
uid:-::::1599120997::F8568D6D3CCC3A3AF2EC0F0E461A6103EB501ECE::Arne Blankerts <arne.blankerts@thephp.cc>::::::::::0:
uid:-::::1599120997::3299E3B0EBAC28CD3122813E9375A0F64C1797DE::Arne Blankerts <arne@thephp.cc>::::::::::0:
sub:-:4096:1:058C154A818434D6:1599120129::::::e::::::23:
fpr:::::::::97ADF2873CF87D544AF916B9058C154A818434D6:
sub:-:4096:1:78CB5411E1CACDB4:1599120165:1756800165:::::a::::::23:
fpr:::::::::AF97A0002978E94B83EC6D3578CB5411E1CACDB4:
sub:-:4096:1:3757412C57E569D3:1599120058::::::s::::::23:
fpr:::::::::D362524427BB163DB089989F3757412C57E569D3:
```

**Comment by heiglandreas** [comment]
Sorry. My mistake. `keys.openpgp.org` of course!

Well. In the case of phive: would we need the UID information? After all it's about verifying that the signature has been done with the key with fingerprint 0xXYZ. 

But the main reason I was raising this issue is this:

```bash
$  phive install phly/keep-a-changelog 
Phive 0.14.4 - Copyright (C) 2015-2020 by Arne Blankerts, Sebastian Heuer and Contributors
Downloading https://api.github.com/rate_limit
Downloading https://github.com/phly/keep-a-changelog/releases/download/2.10.0/keep-a-changelog.phar
Downloading https://github.com/phly/keep-a-changelog/releases/download/2.10.0/keep-a-changelog.phar.asc
Downloading key 9A2577FF9A688FAF
Trying to connect to keys.openpgp.org (37.218.245.50)
Downloading https://keys.openpgp.org/pks/lookup?op=get&options=mr&search=0x9A2577FF9A688FAF
Successfully downloaded key.
[WARNING]  Parsing key data failed with error code 0: No UIDs in key found 
Trying to connect to keyserver.ubuntu.com (162.213.33.9)
Trying to connect to keyserver.ubuntu.com (162.213.33.8)
[WARNING]  Failed with error code 404: Key not found on keyserver 
Trying to connect to hkps.pool.sks-keyservers.net (209.244.105.201)
[WARNING]  Failed with error code 404: Key not found on keyserver 
[ERROR]    PublicKey 9A2577FF9A688FAF not found on key servers 
```

Here a key is downloaded from keys.openpgp.org but on importing the key into the keychain a problem occurs. If the key is – for whatever reason – not available anywhere else, the installation fails.

This is based on a brand new install of GPG in version 2.2.23

So does it make sense to have `keys.openpgp.org` in the list of keyservers to test when the possibly found key can not be imported (and that problem might persist for some time as the GnuPG folks and the OpenPGP folks seem to clash here)?

**Comment by theseer** [comment]
Well, that key is just broken:

```
theseer@nyda /tmp/x5 $ gpg --homedir /tmp/x5 --quiet --with-colons --no-permission-warning --import 0x9A2577FF9A688FAF.asc
gpg: key 9A2577FF9A688FAF: new key but contains no user ID - skipped
```

I don't know where that particular key might have been published, but it's obviously none of the known keyservers. As it seems, `keys.opengpg.org` returns something no matter what key-id you give it, even for keys it doesn't actually know anything about. I so far don't know how to stop it from doing that..


To use a key for signature verification, it needs to get imported into your keyring and by that applying some level of trust... And, for hopefully obvious reasons, the user has to confirm the import of the key - which requires a fingerprint and a UID to be human readable. Without it, it's as bad as openssl keys without any means to verify *who* owns the key.

**Comment by theseer** [comment]
This seems intentional. On `https://keys.openpgp.org/about` is written, they do not publish UIDs for unverified keys. Which makes a lot of sense, actually. But it also means that the key mentioned above is not verified and thus not usable.

I don't see anything that can be done here - except maybe enhancing the error message ;)

**Comment by heiglandreas** [comment]
Yeah. But as far as I see it phive uses that server to check for the key. Which can then not be used. So does it make sense to use that server in the first place?

**Comment by heiglandreas** [comment]
My main concern was that phive uses a keyserver that – by it's own definition – can not be used to import and verify a key. In that case it might make more sense to just remove that keyserver...

**Comment by theseer** [comment]
Well, not exacxtly ;-) 

For keys to useful, they have to have UIDs included. But the "legacy" keyservers allow(ed) to register keys for any email without any verification. So I could create a key with your email address and upload it to the keyservers. While keyservers traditionally make claim that a public key is trustworthy - they just provide a means to transport the information and make it available - this is becoming more and more of a burden.

Among other problems with the legacy servers, this is a key issue address by the next generation key servers. One major change provided by the new keyserver (`keys.openpgp.org`) is that they require uploaded keys to verified, as in, you have to click on a link they send to every UID listed. Otherwise the UID is not included in the response. 

That means, if the key can be found on `keys.openpgp.org` that 2nd step has not been performed (yet). Which effectively makes the key received useless. As stated before, I didn't a way to only ask for verified keys.

Using `keys.openpgp.org` though as one of the sources makes a lot of sense: *If* they key is fully validated and verified, it's more trustworthy than when found on the legacy servers. Many new keys are also only exclusively uploaded to this server as the old key server infrastructure is not exactly gaining in popularity and trustworthiness. (If you have some spare time, google for sks-keyserver security issues ;) )

tl;dr: `keys.openpgp.org` is the future. Make sure you verify your keys and all is well :)
